// ngx-soundmanager2/esm2015/src/ngx-soundmanager2.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ClearPlaylistDirective } from './clear-playlist.directive';
import { DecreaseVolumeDirective } from './decrease-volume.directive';
import { IncreaseVolumeDirective } from './increase-volume.directive';
import { MusicPlayerDirective } from './music-player.directive';
import { MuteMusicDirective } from './mute-music.directive';
import { NextTrackDirective } from './next-track.directive';
import { PauseMusicDirective } from './pause-music.directive';
import { PlayAllDirective } from './play-all.directive';
import { PlayFromPlaylistDirective } from './play-from-playlist.directive';
import { PlayMusicDirective } from './play-music.directive';
import { PreviousTrackDirective } from './previous-track.directive';
import { RemoveFromPlaylistDirective } from './remove-from-playlist.directive';
import { RepeatMusicDirective } from './repeat-music.directive';
import { StopMusicDirective } from './stop-music.directive';
import { HumanTimePipe } from './human-time.pipe';
import { MusicPlayerService } from './music-player.service';
export class NgxSoundmanager2Module {
    constructor() { }
    /**
     * @param {?=} setupOptions
     * @return {?}
     */
    static forRoot(setupOptions) {
        return {
            ngModule: NgxSoundmanager2Module,
            providers: [
                MusicPlayerService,
                { provide: 'setupOptions', useValue: setupOptions }
            ]
        };
    }
}
NgxSoundmanager2Module.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ClearPlaylistDirective,
                    DecreaseVolumeDirective,
                    IncreaseVolumeDirective,
                    MusicPlayerDirective,
                    MuteMusicDirective,
                    NextTrackDirective,
                    PauseMusicDirective,
                    PlayAllDirective,
                    PlayFromPlaylistDirective,
                    PlayMusicDirective,
                    PreviousTrackDirective,
                    RemoveFromPlaylistDirective,
                    RepeatMusicDirective,
                    StopMusicDirective,
                    HumanTimePipe
                ],
                exports: [
                    ClearPlaylistDirective,
                    DecreaseVolumeDirective,
                    IncreaseVolumeDirective,
                    MusicPlayerDirective,
                    MuteMusicDirective,
                    NextTrackDirective,
                    PauseMusicDirective,
                    PlayAllDirective,
                    PlayFromPlaylistDirective,
                    PlayMusicDirective,
                    PreviousTrackDirective,
                    RemoveFromPlaylistDirective,
                    RepeatMusicDirective,
                    StopMusicDirective,
                    HumanTimePipe
                ]
            },] }
];
/** @nocollapse */
NgxSoundmanager2Module.ctorParameters = () => [];
function NgxSoundmanager2Module_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxSoundmanager2Module.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxSoundmanager2Module.ctorParameters;
}
