// ngx-soundmanager2/esm5/src/ngx-soundmanager2.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// Constants
export { MusicPlayerEventConstants } from './music-player-events.constants';
// Directives
export { ClearPlaylistDirective } from './clear-playlist.directive';
export { DecreaseVolumeDirective } from './decrease-volume.directive';
export { IncreaseVolumeDirective } from './increase-volume.directive';
export { MusicPlayerDirective } from './music-player.directive';
export { MuteMusicDirective } from './mute-music.directive';
export { NextTrackDirective } from './next-track.directive';
export { PauseMusicDirective } from './pause-music.directive';
export { PlayAllDirective } from './play-all.directive';
export { PlayFromPlaylistDirective } from './play-from-playlist.directive';
export { PlayMusicDirective } from './play-music.directive';
export { PreviousTrackDirective } from './previous-track.directive';
export { RemoveFromPlaylistDirective } from './remove-from-playlist.directive';
export { RepeatMusicDirective } from './repeat-music.directive';
export { StopMusicDirective } from './stop-music.directive';
// Interfaces
// export * from './itrack-event.interface';
// Module
export { NgxSoundmanager2Module } from './ngx-soundmanager2.module';
// Pipes
export { HumanTimePipe } from './human-time.pipe';
// Services
export { MusicPlayerService } from './music-player.service';
